File: Heap/Q2. Find K Pairs with Smallest Sums/kSmallestPairs.cpp
#include <vector>
#include <priority_queue>

using namespace std;

class Solution {
public:
    vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        //TYPE = [int(sum), int(i), int(j)] => vector<int>
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>>min;
        vector<vector<int>> res;

        for(int i=0; i<nums1.size(); i++){
            min.push({nums1[i]+nums2[0] , i , 0});
        }

        while(!min.empty()){
            vector<int> item = min.top();
            int i = item[1];
            int j = item[2];
            res.push_back({nums1[i], nums2[j]});
            min.pop();
            if(res.size() == k) return res;

            if(j+1 < nums2.size()){
                min.push({nums1[i]+ nums2[j+1] , i , j+1});
            }
        }

        return res;
        
    }
};

File: Heap/Q2. Find K Pairs with Smallest Sums/kSmallestPairs.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number} k
 * @return {number[][]}
 */
var kSmallestPairs = function (nums1, nums2, k) {
    const result = [];
    const heap = new MinPriorityQueue({ compare: (a, b) => a[0] - b[0] });

    for (let i = 0; i < nums1.length; i++) {
        //              sum             , i , j
        heap.enqueue([nums1[i] + nums2[0], i, 0]);
    }
    while (!heap.isEmpty()) {
        const [sum, i, j] = heap.dequeue();
        result.push([nums1[i], nums2[j]]);

        if (result.length === k) return result

        if (j + 1 < nums2.length) {
            heap.enqueue([nums1[i] + nums2[j + 1], i, j + 1]);
        }
    }

    return result;
};